### DISCUSSION TITLE
Can't get Debugging to work

### DISCUSSION THREAD

**Discussion by karlscholz** [bitcraze/discussions#331] [Q&A] [answered]
Hi,
I already did a bit of developing on the crazyflie and switched to WSL for comfort but now I have the need to debug my code on chip. I already gave up on the idea of debugging via WSL and switched over to the official VM, hoping that everything would work out of the box. I followed the guide on the wiki again, but I get this error:
![grafik](https://user-images.githubusercontent.com/105341601/196682645-6c63973c-35c8-42ff-bccb-a99c3f81b61a.png)
in terminal and this in the debug console:
![grafik](https://user-images.githubusercontent.com/105341601/196683126-5f7e23ba-ac26-4243-ac3e-f0c312b21436.png)
I have no idea what I am still doing wrong. Tried everything i found by googling which resulted in me bricking the VM and starting all over again. I was able to successfully connect to the STM via the ST-Link Utility application on windows.
I'm happy for every hint. Thanks!

**Comment by karlscholz** [comment]
afterburner: forgot to show my launch.json file, it looks like this:
```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "STM32 Debug",
            "cwd": "${workspaceRoot}",
            "executable": "${workspaceRoot}/build/cf2.elf",
            "request": "launch",
            "type": "cortex-debug",
            "device": "STM32F405",
            "svdFile": "STM32F405.svd",
            "servertype": "openocd",
            "configFiles": ["interface/stlink-v2.cfg", "target/stm32f4x.cfg"],
            "runToEntryPoint": "main",
            "showDevDebugOutput": "raw",
            //legacy: "runToMain": true,
            "preLaunchCommands": [
                "set mem inaccessible-by-default off",
                "enable breakpoint",
                "monitor reset"
            ]
        },
        // {
        //     "name": "STM32 App Debug"
        // }
    ]
}
```

**Comment by knmcguire** [comment]
Debugging from the stm is perhaps not really advised... especially if you have performance issues. The USB readout is probably not fast enough.

You could perhaps try the USBipd project to connect it to WSL on windows? Or if you perhaps have a native ubuntu machine somewhere?

**Reply by karlscholz** [reply]
I tried the USBipd with WSL, but that sadly didn't work. Not even the st-util can detect the ST Link, even with it being visible under lsusb. WSL was the first i tried, because i like it the most, but since that didn't work i thought i had a better shot with the official VM.
I tried the GDB CLI interface and debugging seemed to work fine with that in the Bitcraze VM:
![grafik](https://user-images.githubusercontent.com/105341601/197473811-cf7a83b1-f099-456a-aff4-97a5027d168d.png)
So the problem must be somewhere in VSCode, Cortex-Debug or the connection between them and GDB.

**Reply by knmcguire** [reply]
Yes it could be... I haven't tried it yet for a while but it might be broken with later versions of the cortex debug.

We are currently in Japan for IROS so we can't test this out right now, but if you could perhaps search the discussions and forums to see if anybody had a similar problem. We hope to be a bit more responsive next week.

**Reply by karlscholz** [reply]
I didn't have the idea of downgrading to a previous cortex debug version yet, i'll give that a try.
Saw it on Insta, no worries. Enjoy your trip! :)

**Reply by karlscholz** [reply]
Tried some legacy versions of Cortex-Debug and it works up to version 1.2.2. 1.4.0 and above currently don't work. With that information the instructions https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/openocd_gdb_debugging/ work. Now i just have to figure out my USB wsl problem... :) Thanks a lot for the hint, I spent way to much time on that.

**Reply by karlscholz** [reply]
![grafik](https://user-images.githubusercontent.com/105341601/197749501-08074aeb-2abf-4175-95e7-53137318bdd3.png)
And there you go, Debugging out of WSL at full USB Speed on Windows 11.
Solution was just a simple chmod:
![grafik](https://user-images.githubusercontent.com/105341601/197750105-fde30e35-5b4b-474b-a2a2-d9ba9d18636d.png)

Would be nice if you could include that Cortex 1.2.2 Version as well as (for WSL) the `sudo chmod +666 /dev/bus/usb/<busid, see lsusb>/<deviceid, see lsusb>` in the getting started with debugging wiki page. Usbipd was straight forward, but here you go:

Windows Powershell(Admin): 
- `winget install --interactive --exact dorssel.usbipd-win`

WSL: 
- `sudo apt install linux-tools-5.4.0-77-generic hwdata sudo update-alternatives --install /usr/local/bin/usbip usbip /usr/lib/linux-tools/5.4.0-77-generic/usbip 20`

Windows Powershell(Admin): 
- `usbipd wsl list`
- `usbipd wsl attach -b <busid of ST-Link>`

**Reply by knmcguire** [reply]
Sorry for the late reply! We were in Japan for some conferences.

and thanks for sharing this! This will be useful for the debugging guide for sure! https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/openocd_gdb_debugging/

Would you be interested in making a PR for the documentation for this? you can use the in browser edit for this directly for github https://github.com/bitcraze/crazyflie-firmware/blob/master/docs/development/openocd_gdb_debugging.md

**Reply by knmcguire** [reply]
Also if you would, perhaps make a brief summary of what worked for you underneath at 'suggest an answer' and then mark it as an answer :)

**Reply by karlscholz** [reply]
Sure, will do the PR later! :)

**Comment by karlscholz** [comment, marked as answer]
# TL:DR: Downgrade your Cortex-Debug Extension to Version 1.2.2
![grafik](https://user-images.githubusercontent.com/105341601/200006974-110b33f3-faa1-4a8e-a231-365e638970a3.png)

# Debugging in WSL requires a bit more: 

## Install usbipd 

Windows Powershell(Admin): 
- `winget install --interactive --exact dorssel.usbipd-win`

WSL: 
- `sudo apt install linux-tools-5.4.0-77-generic hwdata sudo update-alternatives --install /usr/local/bin/usbip usbip /usr/lib/linux-tools/5.4.0-77-generic/usbip 20`

Windows Powershell(Admin): 
- `usbipd wsl list`
- `usbipd wsl attach -b <busid of ST-Link>`

## chmod the usb device, so cortex debug can access it without sudo rights
`sudo chmod +666 /dev/bus/usb/<busid, see lsusb>/<deviceid, see lsusb>`

![grafik](https://user-images.githubusercontent.com/105341601/197750105-fde30e35-5b4b-474b-a2a2-d9ba9d18636d.png)
